### PR TITLE
feat(python): add tag parameter to Branch.transact() API

### PR DESCRIPTION
**Pull request by skshetry**
### Background

This PR addresses [GitHub issue #8200](https://github.com/treeverse/lakeFS/issues/8200) where users requested the ability to create tags automatically during transactions, eliminating the need to manually retrieve commit SHA and create tags separately.

### Bug Fix

N/A - This is a new feature.

### New Feature

Added optional `tag` parameter to `Branch.transact()` method for automatic tag creation after successful transaction completion.

**Usage:**
```python
with branch.transact(commit_message='update', tag='v1.0.0') as tx:
    tx.object('file.txt').upload('data')
# Tag 'v1.0.0' created automatically on merge commit
```

**Key behaviors:**
- Tag created only on transaction success
- Tag accessible via `tx.tag` attribute during transaction

### Testing Details

- All integration tests pass (79/79, excluding known unrelated failure)
- New integration test validates tag functionality
- Perfect pylint scores (10.00/10)
- All existing functionality remains unchanged

### Breaking Change?

No. This is a backward-compatible addition - existing transaction code works unchanged.

## Additional info

**Files modified:**
- `lakefs/branch.py`: Added tag parameter and creation logic
- `tests/integration/test_branch.py`: Added tag functionality test

### Contact Details

Available via GitHub for questions.

### CODE REVIEW

File: clients/python-wrapper/lakefs/branch.py [hunk 1]
```diff
@@ -417,11 +437,15 @@ def __exit__(self, typ, value, traceback) -> bool:
...
-
-            return False
         except LakeFSException as e:
             if self._cleanup_branch:
                 self._tx_branch.delete()
             raise TransactionException(f"Failed committing transaction {self._tx.id}: {e}") from e
+
+        if self._tx.tag:
+            tag = Tag(self._repo_id, self._tx.tag, self._client)
+            tag.create(merge_commit_id)
```
Comment (skshetry) [inline clients/python-wrapper/lakefs/branch.py:450]: A few open questions here:

1. What should happen if the tag already exists? Should a new tag be force-created?
2. What should happen if the tag creation fails? Does the transaction needs to be reverted? I don't think we should but just asking.

File: clients/python-wrapper/lakefs/branch.py [hunk 2]
```diff
@@ -382,6 +385,21 @@ def commit_metadata(self, metadata: Optional[Dict]) -> None:
...
 
+    @property
+    def tag(self) -> Optional[str]:
+        """
+        Return the tag name configured for this transaction completion
+        """
+        return self._tag
+
+    @tag.setter
+    def tag(self, tag: Optional[str]) -> None:
```
Comment (N-o-Z) [inline clients/python-wrapper/lakefs/branch.py]: Should not be optional on setter (unless we want to unset tag?? which kind of makes this setter behavior strange

File: clients/python-wrapper/lakefs/branch.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:450]: In case tag creation fails (exists, other failure) this should not fail the transaction - a tag could be added manually at any given point and it does not impact the integrity of the transaction

File: clients/python-wrapper/tests/integration/test_branch.py [hunk 3]
```diff
@@ -207,3 +207,14 @@ def test_transaction_failure(setup_repo, cleanup_branch):
         assert tx.get_commit()
 
     assert test_branch.get_commit() == new_ref.get_commit()
+
+
+def test_transaction_with_tag(setup_repo):
```
Comment (N-o-Z) [inline clients/python-wrapper/tests/integration/test_branch.py:214]: 1. Add a check in the old transaction test case that lists tags and verifies it is an empty list.
2. Add a test passing explicit None in tag and very list tags is empty

File: clients/python-wrapper/lakefs/branch.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:450]: To clarify - tag creation failure should not require revert of transaction and commit and merge should be successful as well as tx_branch cleanup if applicable. We should somehow notify the user on the failure

Comment (N-o-Z) [reply clients/python-wrapper/lakefs/branch.py:450]: I think the location you added the code is good

File: clients/python-wrapper/tests/integration/test_branch.py (hunk 3 above)
Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_branch.py:214]: 3. Add a test where tagging fails (already exists) and ensure:
- Transaction completed successfully
- tx_branch deleted

Comment (skshetry) [reply clients/python-wrapper/tests/integration/test_branch.py:214]: I have added test cases that you suggested. PTAL.

File: clients/python-wrapper/tests/integration/test_branch.py [hunk 4]
```diff
@@ -207,3 +209,45 @@ def test_transaction_failure(setup_repo, cleanup_branch):
...
+    path_and_data = ["a", "b", "bar/a", "bar/b", "bar/c", "c"]
+    test_branch = repo.branch("main")
+
+    repo: "lakefs.Repository"
+
+    test_branch.object("initial_file").upload("initial content")
+    initial_commit = test_branch.commit("initial commit")
+    repo.tag("v1.0.0").create(initial_commit)
+
+    with expect_exception_context(ConflictException, "tag already exists"):
```
Comment (skshetry) [inline clients/python-wrapper/tests/integration/test_branch.py:247]: I wonder why we use this custom function instead of using [`pytest.raises(..., match=...)`](https://docs.pytest.org/en/stable/reference/reference.html#pytest-raises).

This is equivalent to:

```suggestion
    with pytest.raises(ConflictException, match="tag already exists"):
```

**Review by itaiad200 — DISMISSED** [review]
wip

File: clients/python-wrapper/tests/integration/test_branch.py [hunk 5]
```diff
@@ -207,3 +209,45 @@ def test_transaction_failure(setup_repo, cleanup_branch):
...
+        with test_branch.transact(commit_message="my transaction with existing tag", tag="v1.0.0") as tx:
+            upload_data(tx, path_and_data)
+
+    # Verify transaction branch was deleted
+    with expect_exception_context(NotFoundException):
+        repo.branch(tx.id).get_commit()
```
Comment (N-o-Z) [inline clients/python-wrapper/tests/integration/test_branch.py:253]: Test should not assume transaction successful.
Need to verify data

**Review by N-o-Z — APPROVED** [review]
LGTM with one comment.
In addition, please refrain from force pushing or rebasing your branch while review is in progress.
It makes the review process longer and more difficult

**Comment by skshetry** [comment]
> LGTM with one comment.
> In addition, please refrain from force pushing or rebasing your branch while review is in progress.
> It makes the review process longer and more difficult

~~If you are not aware, you may find the 'Compare' button above on GitHub useful for reviewing these changes.~~

Sorry, it looks like GitHub hides previous pushes nowadays (seems to have happened somewhere around mid 2022), and only shows you latest pushes. 🤦🏽 

- https://github.com/orgs/community/discussions/3478

File: clients/python-wrapper/tests/integration/test_branch.py (hunk 5 above)
Comment (skshetry) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: This was supposed to be just a sanity check, as some other tests likely verify the data.

Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: There are no old tests that verify the data upon failure to tag the commit

Comment (skshetry) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: Failure to create a tag does not fail the transaction as it happens after the fact.

Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: This is you speaking as a developer who read the code.
The test is not supposed to assume that.
If we assumed our code does exactly what we want we wouldn't be writing tests.

Comment (skshetry) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: Would you be okay with testing that the `branch` moved to a new commit (i.e is not the `initial_commit`)? Would that be enough to verify the transaction was successful? 

Or, do you think we should verify the data?

Comment (N-o-Z) [reply clients/python-wrapper/tests/integration/test_branch.py:253]: I think checking the commit should be sufficient

**Comment by skshetry** [comment]
Hey @N-o-Z, sorry about leaving this hanging, without any updates. 
I'll address https://github.com/treeverse/lakeFS/pull/9334#discussion_r2261082901 this week.
